pkg/agent/packetcapture/packetcapture_controller_linux.go

```diff
@@ -0,0 +1,884 @@
+		}()
+		c.queue.Add(capture)
+	}
+}
+
+func (c *Controller) processPacketCaptureItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+	if err := c.syncPacketCapture(key); err == nil {
+		c.queue.Forget(key)
+	} else {
+		c.queue.AddRateLimited(key)
+		klog.ErrorS(err, "Error syncing PacketCapture, requeueing", "key", key)
+	}
+	return true
+}
+
+func (c *Controller) cleanupPacketCapture(pcName string) {
+	path := nameToPath(pcName)
+	exist, err := afero.Exists(defaultFS, path)
+	if err != nil {
+		klog.ErrorS(err, "Failed to check if path exists", "path", path)
+	}
+	if !exist {
+		return
+	}
+	if err := defaultFS.Remove(path); err == nil {
+		klog.V(2).InfoS("Deleted pcap file", "name", pcName, "path", path)
+	} else {
+		klog.ErrorS(err, "Failed to delete pcap file", "name", pcName, "path", path)
+	}
+}
+
+func getPacketFileAndWriter(name string) (afero.File, *pcapgo.NgWriter, error) {
+	filePath := nameToPath(name)
+	var file afero.File
+	if _, err := os.Stat(filePath); err == nil {
+		return nil, nil, fmt.Errorf("packet file already exists. this may be due to an unexpected termination")
+	} else if os.IsNotExist(err) {
+		file, err = defaultFS.Create(filePath)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to create pcapng file: %w", err)
+		}
+	} else {
+		return nil, nil, fmt.Errorf("couldn't check if the file exists: %w", err)
+	}
+	writer, err := pcapgo.NewNgWriter(file, layers.LinkTypeEthernet)
+	if err != nil {
+		return nil, nil, fmt.Errorf("couldn't initialize pcap writer: %w", err)
+	}
+	return file, writer, nil
+}
+
+// getTargetCaptureDevice is trying to locate the target device for packet capture. If the target Pod does not exist on the current Node,
+// the agent on this Node will not perform the capture.
+func (c *Controller) getTargetCaptureDevice(pc *crdv1alpha1.PacketCapture) *string {
+	var pod, ns string
+	if pc.Spec.Source.Pod != nil {
+		pod = pc.Spec.Source.Pod.Name
+		ns = pc.Spec.Source.Pod.Namespace
+	} else {
+		pod = pc.Spec.Destination.Pod.Name
+		ns = pc.Spec.Destination.Pod.Namespace
+	}
+
+	podInterfaces := c.interfaceStore.GetContainerInterfacesByPod(pod, ns)
+	if len(podInterfaces) == 0 {
+		return nil
+	}
+
+	return &podInterfaces[0].InterfaceName
+}
+
+func (c *Controller) startPacketCapture(name string) {
+	klog.V(4).InfoS("Started processing PacketCapture", "name", name)
+
+	go func() error {
+		var err error
+		pcState := &packetCaptureState{name: name}
+		defer func() {
+			if err != nil {
+				c.cleanupPacketCapture(name)
+				klog.ErrorS(err, "PackageCapture failed", "name", name)
+				err := c.updatePacketCaptureStatus(name, pcState.capturedPacketsNum, "", err)
+				if err != nil {
+					klog.ErrorS(err, "failed to update PacketCapture status")
+				}
+			}
+		}()
+		pc, err := c.packetCaptureLister.Get(name)
+		if err != nil {
+			return err
+		}
+
+		device := c.getTargetCaptureDevice(pc)
+		if device == nil {
+			return nil
+		}
```

Comment:
This is duplicated with the device check in the following `c.newPacketSourceFn`, I think this can be removed.

Reply:
@I feel like this check happens too late. If the capture is not local, I don't think we should even call `startPacketCapture`. We should do the check early in `syncPacketCapture` and return immediately if the capture is not local. Am I missing something?

Reply:
updated.